## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to import the movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [1]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [27]:
# Write your code for importing the csv file here
movies = pd.read_csv("/content/sample_data/Movies.csv")
movies

,color,director_name,num_critic_for_reviews,duration,director_facebook_likes,actor_3_facebook_likes,actor_2_name,actor_1_facebook_likes,gross,genres,...,num_user_for_reviews,language,country,content_rating,budget,title_year,actor_2_facebook_likes,imdb_score,aspect_ratio,movie_facebook_likes
0,Color,James Cameron,723.0,178.0,0.0,855.0,Joel David Moore,1000.0,760505847.0,Action|Adventure|Fantasy|Sci-Fi,...,3054.0,English,USA,PG-13,237000000.0,2009.0,936.0,7.9,1.78,33000
1,Color,Gore Verbinski,302.0,169.0,563.0,1000.0,Orlando Bloom,40000.0,309404152.0,Action|Adventure|Fantasy,...,1238.0,English,USA,PG-13,300000000.0,2007.0,5000.0,7.1,2.35,0
2,Color,Sam Mendes,602.0,148.0,0.0,161.0,Rory Kinnear,11000.0,200074175.0,Action|Adventure|Thriller,...,994.0,English,UK,PG-13,245000000.0,2015.0,393.0,6.8,2.35,85000
3,Color,Christopher Nolan,813.0,164.0,22000.0,23000.0,Christian Bale,27000.0,448130642.0,Action|Thriller,...,2701.0,English,USA,PG-13,250000000.0,2012.0,23000.0,8.5,2.35,164000
4,Color,Andrew Stanton,462.0,132.0,475.0,530.0,Samantha Morton,640.0,73058679.0,Action|Adventure|Sci-Fi,...,738.0,English,USA,PG-13,263700000.0,2012.0,632.0,6.6,2.35,24000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3848,Color,Shane Carruth,143.0,77.0,291.0,8.0,David Sullivan,291.0,424760.0,Drama|Sci-Fi|Thriller,...,371.0,English,USA,PG-13,7000.0,2004.0,45.0,7.0,1.85,19000
3849,Color,Neill Dela Llana,35.0,80.0,0.0,0.0,Edgar Tancangco,0.0,70071.0,Thriller,...,35.0,English,Philippines,Not Rated,7000.0,2005.0,0.0,6.3,NaN,74
3850,Color,Robert Rodriguez,56.0,81.0,0.0,6.0,Peter Marquardt,121.0,2040920.0,Action|Crime|Drama|Romance|Thriller,...,130.0,Spanish,USA,R,7000.0,1992.0,20.0,6.9,1.37,0
3851,Color,Edward Burns,14.0,95.0,0.0,133.0,Caitlin FitzGerald,296.0,4584.0,Comedy|Drama,...,14.0,English,USA,Not Rated,9000.0,2011.0,205.0,6.4,NaN,413


**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [28]:
# Write your code for inspection here
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3853 entries, 0 to 3852
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      3851 non-null   object 
 1   director_name              3853 non-null   object 
 2   num_critic_for_reviews     3852 non-null   float64
 3   duration                   3852 non-null   float64
 4   director_facebook_likes    3853 non-null   float64
 5   actor_3_facebook_likes     3847 non-null   float64
 6   actor_2_name               3852 non-null   object 
 7   actor_1_facebook_likes     3853 non-null   float64
 8   gross                      3853 non-null   float64
 9   genres                     3853 non-null   object 
 10  actor_1_name               3853 non-null   object 
 11  movie_title                3853 non-null   object 
 12  num_voted_users            3853 non-null   int64  
 13  cast_total_facebook_likes  3853 non-null   int64

#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='red'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

In [11]:
# Write you code here

movies.isna().sum()

,0
color,2
director_name,0
num_critic_for_reviews,1
duration,1
director_facebook_likes,0
actor_3_facebook_likes,6
actor_2_name,1
actor_1_facebook_likes,0
gross,0
genres,0


#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

In [ ]:
# Write you code here



### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  `color`
-  `director_facebook_likes`
-  `actor_1_facebook_likes`
-  `actor_2_facebook_likes`
-  `actor_3_facebook_likes`
-  `actor_2_name`
-  `cast_total_facebook_likes`
-  `actor_3_name`
-  `duration`
-  `facenumber_in_poster`
-  `content_rating`
-  `country`
-  `movie_imdb_link`
-  `aspect_ratio`
-  `plot_keywords`

In [29]:
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations

clean_movies=movies.drop(columns=['color', 'director_facebook_likes', 'actor_1_facebook_likes', 'actor_2_facebook_likes',
                     'actor_3_facebook_likes', 'actor_2_name', 'cast_total_facebook_likes', 'actor_3_name',
                     'duration', 'facenumber_in_poster', 'content_rating', 'country', 'movie_imdb_link',
                     'aspect_ratio', 'plot_keywords'])

#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='red'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'.

In [15]:
# Write you code here
clean_movies.isna().sum()


,0
director_name,0
num_critic_for_reviews,1
gross,0
genres,0
actor_1_name,0
movie_title,0
num_voted_users,0
num_user_for_reviews,0
language,4
budget,0


#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [20]:
clean_movies['language'].fillna('English', inplace=True)

/tmp/ipykernel_498/3027005558.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  clean_movies['language'].fillna('English', inplace=True)


In [19]:
english_movies_count = clean_movies[clean_movies['language'] == 'English'].shape[0]
print(f"Number of movies in English: {english_movies_count}")

Number of movies in English: 3675


#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [30]:
# Write your code for unit conversion here
# Method 2: Create a formatted string column with '$' and 'M' (e.g., $5.00M)
clean_movies['budget'] = (clean_movies['budget'] / 1000000)
clean_movies['gross'] = (clean_movies['gross'] / 1000000)



**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [32]:
# Write your code for creating the profit column here
#clean_movies.head()
clean_movies['profit'] = clean_movies['gross'] - clean_movies['budget']

In [34]:
# Write your code for sorting the dataframe here
clean_movies=clean_movies.sort_values('profit', ascending=False)


,director_name,num_critic_for_reviews,gross,genres,actor_1_name,movie_title,num_voted_users,num_user_for_reviews,language,budget,title_year,imdb_score,movie_facebook_likes,profit
0,James Cameron,723.0,760.505847,Action|Adventure|Fantasy|Sci-Fi,CCH Pounder,Avatar,886204,3054.0,English,237.0,2009.0,7.9,33000,523.505847
28,Colin Trevorrow,644.0,652.177271,Action|Adventure|Sci-Fi|Thriller,Bryce Dallas Howard,Jurassic World,418214,1290.0,English,150.0,2015.0,7.0,150000,502.177271
25,James Cameron,315.0,658.672302,Drama|Romance,Leonardo DiCaprio,Titanic,793059,2528.0,English,200.0,1997.0,7.7,26000,458.672302
2704,George Lucas,282.0,460.935665,Action|Adventure|Fantasy|Sci-Fi,Harrison Ford,Star Wars: Episode IV - A New Hope,911097,1470.0,English,11.0,1977.0,8.7,33000,449.935665
2748,Steven Spielberg,215.0,434.949459,Family|Sci-Fi,Henry Thomas,E.T. the Extra-Terrestrial,281842,515.0,English,10.5,1982.0,7.9,34000,424.449459
16,Joss Whedon,703.0,623.279547,Action|Adventure|Sci-Fi,Chris Hemsworth,The Avengers,995415,1722.0,English,220.0,2012.0,8.1,123000,403.279547
482,Roger Allers,186.0,422.783777,Adventure|Animation|Drama|Family|Musical,Matthew Broderick,The Lion King,644348,656.0,English,45.0,1994.0,8.5,17000,377.783777
230,George Lucas,320.0,474.544677,Action|Adventure|Fantasy|Sci-Fi,Natalie Portman,Star Wars: Episode I - The Phantom Menace,534658,3597.0,English,115.0,1999.0,6.5,13000,359.544677
64,Christopher Nolan,645.0,533.316061,Action|Crime|Drama|Thriller,Christian Bale,The Dark Knight,1676169,4667.0,English,185.0,2008.0,9.0,37000,348.316061
419,Gary Ross,673.0,407.999255,Adventure|Drama|Sci-Fi|Thriller,Jennifer Lawrence,The Hunger Games,701607,1959.0,English,78.0,2012.0,7.3,140000,329.999255


In [40]:
top10 = clean_movies.head(10)
top10['movie_title']

,movie_title
0,Avatar
28,Jurassic World
25,Titanic
2704,Star Wars: Episode IV - A New Hope
2748,E.T. the Extra-Terrestrial
16,The Avengers
482,The Lion King
230,Star Wars: Episode I - The Phantom Menace
64,The Dark Knight
419,The Hunger Games


**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

**Subtask 3.3: Find IMDb Top 250**

Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000.

Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [41]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe
# and name that dataframe as 'IMDb_Top_250'
IMDb_Top_250 = clean_movies[clean_movies['num_voted_users'] > 25000].sort_values('imdb_score', ascending=False).head(250)
IMDb_Top_250['Rank'] = range(1, 251)


#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_250*? </font>

In [42]:
bins = [7.5, 8.0, 8.5, 9.0, 9.5, 10.0]
labels = ['7.5 to 8', '8 to 8.5', '8.5 to 9', '9 to 9.5', '9.5 to 10']

IMDb_Top_250['imdb_bucket'] = pd.cut(IMDb_Top_250['imdb_score'], bins=bins, labels=labels, right=False)
bucket_counts = IMDb_Top_250['imdb_bucket'].value_counts().sort_index()

print("Number of movies in each IMDb score bucket:")
print(bucket_counts)

max_bucket = bucket_counts.idxmax()
max_count = bucket_counts.max()

print(f"\nThe bucket with the maximum number of movies is '{max_bucket}' with {max_count} movies.")

Number of movies in each IMDb score bucket:
imdb_bucket
7.5 to 8      45
8 to 8.5     159
8.5 to 9      42
9 to 9.5       4
9.5 to 10      0
Name: count, dtype: int64

The bucket with the maximum number of movies is '8 to 8.5' with 159 movies.


**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [43]:
# Write your code for creating three new dataframes here
Meryl_Streep = IMDb_Top_250[IMDb_Top_250['actor_1_name'] == 'Meryl Streep']

In [44]:
Leo_Caprio = IMDb_Top_250[IMDb_Top_250['actor_1_name'] == 'Leonardo DiCaprio']

In [46]:
Brad_Pitt = IMDb_Top_250[IMDb_Top_250['actor_1_name'] == 'Brad Pitt']

In [57]:
# Write your code for combining the three dataframes here
Combined = pd.concat([Meryl_Streep, Brad_Pitt, Leo_Caprio], axis=0, ignore_index=True)

In [58]:
# Write your code for grouping the combined dataframe here
Combined=Combined.groupby(by='actor_1_name')


In [61]:
# Write the code for finding the mean of critic reviews and audience reviews here
Combined.describe()

num_critic_for_reviews                                 \
                                   count        mean         std    min   
actor_1_name                                                              
Brad Pitt                            2.0  218.500000  136.471609  122.0   
Leonardo DiCaprio                    8.0  471.375000  215.606081  166.0   
Meryl Streep                        11.0  181.454545   89.673144   42.0   

                                                gross              ...  \
                      25%    50%     75%    max count        mean  ...   
actor_1_name                                                       ...   
Brad Pitt          170.25  218.5  266.75  315.0   2.0   24.652448  ...   
Leonardo DiCaprio  312.50  523.0  615.00  765.0   8.0  154.752435  ...   
Meryl Streep       120.00  208.0  230.50  331.0  11.0   59.919727  ...   

                      profit              Rank                             \
                         75%         max count     mean         std   min   
actor_1_name                                                                
Brad Pitt          -7.033026   -0.718500   2.0   92.500  115.258405  11.0   
Leonardo DiCaprio  75.212291  132.568851   8.0  100.375   64.389301  12.0   
Meryl Streep       43.830719   89.732962   0.0      NaN         NaN   NaN   

                                                
                     25%    50%     75%    max  
actor_1_name                                    
Brad Pitt          51.75   92.5  133.25  174.0  
Leonardo DiCaprio  44.00  105.5  145.25  187.0  
Meryl Streep         NaN    NaN     NaN    NaN  

[3 rows x 80 columns]

#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>